src/models/ctx/update_profile.rs

```diff
@@ -101,54 +101,37 @@ pub fn update_profile<E: Env + 'static>(
-                        .map(|addon| &addon.manifest.id)
-                        .enumerate()
-                        .filter(|(_, id)| **id == addon.manifest.id && !addon.flags.protected)
-                        .map(|(position, _)| position)
-                        .collect::<Vec<_>>();
-                    for position in addon_positions {
-                        profile.addons.remove(position);
-                    }
-                    profile.addons.push(addon.to_owned());
-                    let push_to_api_effects = match profile.auth_key() {
-                        Some(auth_key) => Effects::one(push_addons_to_api::<E>(
-                            profile.addons.to_owned(),
-                            auth_key,
-                        ))
-                        .unchanged(),
-                        _ => Effects::none().unchanged(),
-                    };
-                    Effects::msg(Msg::Event(Event::AddonUpgraded {
-                        transport_url: addon.transport_url.to_owned(),
-                        id: addon.manifest.id.to_owned(),
-                    }))
-                    .join(push_to_api_effects)
-                    .join(Effects::msg(Msg::Internal(Internal::ProfileChanged)))
-                } else {
-                    Effects::msg(Msg::Event(Event::Error {
-                        error: CtxError::from(OtherError::AddonConfigurationRequired),
-                        source: Box::new(Event::AddonUpgraded {
-                            transport_url: addon.transport_url.to_owned(),
-                            id: addon.manifest.id.to_owned(),
-                        }),
-                    }))
-                    .unchanged()
-                }
-            } else {
-                Effects::msg(Msg::Event(Event::Error {
-                    error: CtxError::from(OtherError::AddonAlreadyInstalled),
-                    source: Box::new(Event::AddonUpgraded {
-                        transport_url: addon.transport_url.to_owned(),
-                        id: addon.manifest.id.to_owned(),
-                    }),
-                }))
-                .unchanged()
+            if profile.addons.contains(addon) {
+                return addon_upgrade_error_effects(addon, OtherError::AddonAlreadyInstalled);
             }
+            if addon.manifest.behavior_hints.configuration_required {
+                return addon_upgrade_error_effects(addon, OtherError::AddonConfigurationRequired);
+            }
+            let addon_position = profile
+                .addons
+                .iter()
+                .map(|addon| &addon.transport_url)
+                .position(|transport_url| *transport_url == addon.transport_url);
+            if addon_position.is_none() {
+                return addon_upgrade_error_effects(addon, OtherError::AddonNotInstalled);
+            }
+            if addon.flags.protected || profile.addons[addon_position.unwrap()].flags.protected {
```

Comment:
```suggestion
            let mut profile_addon = match profile
                .addons
                .iter_mut()
                .find(|profile_addon| *transport_url == profile_addon.transport_url) {
                    Some(profile_addon) => profile_addon,
                    None => return addon_upgrade_error_effects(addon, OtherError::AddonNotInstalled),
                };

            if addon.flags.protected || profile_addon.flags.protected {
```

it might not be possible to replace the full addon in the profile when using `iter_mut`   and `*profile_addon = addon` but you can try anyway.

Reply:
I wasn't able to replace the full object, but updated to use matching to find the position instead.